### DISCUSSION TITLE
Elsa 2 and .NET 5.0

### DISCUSSION THREAD

**Discussion by mohdali** [elsa-workflows/elsa-core#1106] [Q&A] [answered]
Hi,

We are currently using Elsa 1 and considering to move to Elsa 2 as we can see lots of improvements in Timer, Dashboard Designer, workflow context etc.

One point we have to address is about .NET dependency. We are currently using .NET Core 3.1 since it is an LTS release and plant to move to .NET 6 directly when it is released since it is also going to be LTS (target November this year).

While Elsa Core is Targeting .NET Standard 2.1, many of the other projects (e.g. EF Core Persistence) are targeting .NET 5. However, EF Core 5 itself is a .NET Standard library and does not required .NET 5.

So my question is whether the dependencies can be re-evaluated to use .NET Standard whenever possible to allow a more flexible upgrade paths.

**Comment by sfmskywalker** [comment, marked as answer]
Hi Mohamed,

We've tried to support .NET Core 3.1 on a couple of occasions, and we can probably make it work but given my limited bandwidth I have to make choices on what aspects I spend my time on.

On a different note, I'm not sure I get why it is important to stay on LTS. As far as I know, ALL .NET releases are supported - including .NET 5 - which is a stable release just like 3.1. Its support is just not as long-term as 3.1 (which is only through December 2022) and 6.0. So the way I see it, you might just as well upgrade to 5.0, enjoy its support while it lasts, and upgrade to 6.0 once that becomes available. Not only will you not lose support at any one time, it will make upgrading a smooth operation too since the divergence will be kept to a minimum from one version to the next.

It's up to you of course, and if you are willing to provide a PR with multi-targeting projects, I will gladly accept.

**Reply by mohdali** [reply]
Hi Sipke,

Thank you for your reply. And really appreciate your work on this project, Elsa 2 is coming out great and it is very impressive.

It is a business requirement to have LTS and upgrades do not normally happen that quickly for our target environment. We also don't want to end up with having multiple runtimes side by side as we have other components and dependencies that don't get upgraded that quickly.

I understand the restrictions and will try to work out if we can generate a PR for multi-targeting.

Thanks a lot for all your efforts.

**Reply by sfmskywalker** [reply]
Thanks Mohamed, that’s very kind.

I understand. And your not alone in that situation of course, so if you are able to send a PR with multi targeting then that would be tremendously helpful to others, that I am sure of.

And as always, let me know if you run into any issues.

Thanks for your support.

**Reply by mohdali** [reply]
Hi @sfmskywalker,

I made an initial attempt at multi targeting. Could you please take a look and give your feedback?

https://github.com/mohdali/elsa-core/commit/75a4984b642c1ff13830a7c4e9ce66b9c21e7719

I have not included any multi targeting on samples or test code to simplify.

I tried to target .NET Standard 2.1 if possible, otherwise I included multi targeting.

For EF Core, I selected  .NET Standard 2.1 , However,  I think some designer packages require a target framework. I'm not sure if you are using those. In that case it can be also changed to target both .NET 3.1 and .NET 5.

Almost all the changes are in the csproj files. But there is only one instance where the code needed to be modified as some feature was not available in .NET Core 3.1.

The code builds and able to generate nuget packages in my local test environment. I will proceed to test with a complete .NET Core 3.1 environment to further check.

However, I wanted to know if any concerns on the approach in general.

**Reply by sfmskywalker** [reply]
Regarding the designer packages, I have a few projects that target a framework, which are:

- Elsa.Persistence.EntityFramework.MySql
- Elsa.Persistence.EntityFramework.PostgreSql
- Elsa.Persistence.EntityFramework.Sqlite
- Elsa.Persistence.EntityFramework.SqlServer

Beyond that, your commit looks good to me. Thanks!
